### PR TITLE
WorkerDOM: support long tasks

### PR DESCRIPTION
**Pull request by dvoytenko**

### CODE REVIEW

File: extensions/amp-script/0.1/user-activation-tracker.js [hunk 1]
```diff
@@ -80,6 +83,28 @@ export class UserActivationTracker {
...
 
+  /**
+   * @param {!Promise} promise
+   */
+  expandLongTask(promise) {
+    if (!this.isActive()) {
+      return;
+    }
+    this.inLongTask_ = true;
+    promise.catch(() => {}).then(() => {
```
Comment (choumx) [inline extensions/amp-script/0.1/user-activation-tracker.js]: `promise.catch(() => {}).then(...)` is interesting. Different behavior vs. `Promise.resolve().then(...)`?

File: extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js [hunk 2]
```diff
@@ -73,4 +73,66 @@ describes.realWin('UserActivationTracker', {}, env => {
...
+
+      clock.tick(ACTIVATION_TIMEOUT + 1);
+      expect(tracker.isActive()).to.be.true;
+
+      clock.tick(ACTIVATION_TIMEOUT + 1);
+      expect(tracker.isActive()).to.be.true;
+
+      return promise.then(() => {
+        // Skip microtask.
+        return Promise.resolve();
```
Comment (choumx) [inline extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js]: Nit: Yield macrotask e.g. `poll(1)` may be less brittle.

File: extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js [hunk 3]
```diff
@@ -73,4 +73,66 @@ describes.realWin('UserActivationTracker', {}, env => {
...
+        expect(tracker.isActive()).to.be.true;
+
+        clock.tick(2);
+        expect(tracker.isActive()).to.be.false;
+      });
+    });
+
+    it('should NOT expand when not active', () => {
+      tracker.expandLongTask(Promise.resolve());
+      expect(tracker.isActive()).to.be.false;
```
Comment (choumx) [inline extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js]: Not sure this assertion fully covers the intended scenario. Maybe we should verify activation time:

```js
const t = tracker.getLastActivationTime();
const promise = Promise.resolve();

expect(tracker.isActive()).to.be.false;
tracker.expandLongTask(promise);
yield promise;

// expandLongTask() is a no-op when inactive.
expect(tracker.getLastActivationTime()).to.equal(t);
expect(tracker.isActive()).to.be.false;
```

File: extensions/amp-script/0.1/user-activation-tracker.js (hunk 1 above)
Comment (dvoytenko) [reply extensions/amp-script/0.1/user-activation-tracker.js]: Yeah. This is basically a DYI `finally()`.

File: extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js (hunk 2 above)
Comment (dvoytenko) [reply extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js]: What's `poll(1)`? Do you mean `delay(1)`?

File: extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js (hunk 3 above)
Comment (dvoytenko) [reply extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js]: It should cover it b/c I sleep for a pre-expiration time and check this:

```
clock.tick(ACTIVATION_TIMEOUT - 1);
expect(tracker.isActive()).to.be.true;
```

I.e. this validates that the active status is extended an extra interval.

File: extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js (hunk 2 above)
Comment (dvoytenko) [reply extensions/amp-script/0.1/test/unit/test-user-activation-tracker.js]: I think I got it. Updated the code.

**Pull request merged**